### ISSUE TITLE
[deeplab] Slow inference speed

### ISSUE THREAD

**Issue by cclo-astri** [tensorflow/models#8464]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **yes**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **18.04 (Docker image 1.14-py3-gpu)**
- Mobile device (e.g., Pixel 4, Samsung Galaxy 10) if the issue happens on mobile device: **N/A**
- TensorFlow installed from (source or binary): **bundled in Docker image 1.14-py3-gpu**
- TensorFlow version (use command below): **v1.14.0-rc1-22-gaf24dc91b5 1.14.0**
- Python version: **3.6.8**
- Bazel version (if compiling from source): **N/A**
- GCC/Compiler version (if compiling from source): **N/A**
- CUDA/cuDNN version: **7.4.1.5-1**
- GPU model and memory: **NVIDIA GTX 1080 (8GB)**

**Please provide the entire URL of the model you are using?**
 [https://github.com/tensorflow/models/blob/master/research/deeplab](url)

**Describe the current behavior**

- A quite slow inference speed. In my i7-7800X + GTX 1080 (8GB RAM) machine, using self-trained DeepLabV3+ (Xception) model, needs around 2.5+ seconds to inference a image. 

- The inference speed does not change no matter I changed the input image size from 512x512 to 256x256. 

- And I also tested with the default PASCAL VOC trained model (http://download.tensorflow.org/models/deeplabv3_pascal_train_aug_2018_01_04.tar.gz), the inference time is nearly the same (3 seconds per image).

- Does my machine (especially the graphics card) is too slow to run DeepLabV3+ (Xception) ? 

**Describe the expected behavior**
I am expecting a sub-second (around 500ms or lower) inference time. 

**Code to reproduce the issue**

**Other info / logs**

**Comment by cclo-astri** [comment]
Hi all,

Any updates ?

Thanks.

**Comment by YknZhu** [comment]
Thanks for your interest in deeplab! Inference time can be measured differently. The [reference model](http://download.tensorflow.org/models/deeplabv3_pascal_train_aug_2018_01_04.tar.gz) should run at ~0.7 sec on CPU, so 3 second is a bit slower than expected. 

Could you provide more details? Specifically:
- is the inference run on GPU? please see https://www.tensorflow.org/guide/gpu for more details to use GPU.
- loading model / graph usually takes long. the inference time should be measured after graph is loaded, and it can be reused to inference multiple images.
- could you check the models you are using are single-scale, no left-right flip are used? multiple scale would increase inference time significantly

regarding reducing image size doesn't reduce inference time: the model takes any input size image and will resize it to crop_size (see export_model.py for details). You need to export a smaller crop_size model for running on low res inputs.

**Comment by cclo-astri** [comment]
Hi,

Below is the output message during inference:

> WARNING: Logging before flag parsing goes to stderr.
W0506 10:02:38.399849 140559569971008 deprecation_wrapper.py:119] From ./inference.py:39: The name tf.GraphDef is deprecated. Please use tf.compat.v1.GraphDef instead.
W0506 10:02:39.140044 140559569971008 deprecation_wrapper.py:119] From ./inference.py:42: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.
2020-05-06 10:02:39.157508: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcuda.so.1
2020-05-06 10:02:39.181505: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.835
pciBusID: 0000:65:00.0
2020-05-06 10:02:39.181854: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-05-06 10:02:39.183680: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2020-05-06 10:02:39.184981: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10.0
2020-05-06 10:02:39.185269: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10.0
2020-05-06 10:02:39.186888: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10.0
2020-05-06 10:02:39.188073: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10.0
2020-05-06 10:02:39.191215: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2020-05-06 10:02:39.192929: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-05-06 10:02:39.193273: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2020-05-06 10:02:39.315412: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x4b13a60 executing computations on platform CUDA. Devices:
2020-05-06 10:02:39.315480: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): GeForce GTX 1080, Compute Capability 6.1
2020-05-06 10:02:39.318540: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 3504310000 Hz
2020-05-06 10:02:39.319157: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x13d1b850 executing computations on platform Host. Devices:
2020-05-06 10:02:39.319201: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2020-05-06 10:02:39.320253: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: 
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.835
pciBusID: 0000:65:00.0
2020-05-06 10:02:39.320307: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-05-06 10:02:39.320330: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
2020-05-06 10:02:39.320351: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10.0
2020-05-06 10:02:39.320372: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10.0
2020-05-06 10:02:39.320392: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10.0
2020-05-06 10:02:39.320412: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10.0
2020-05-06 10:02:39.320434: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2020-05-06 10:02:39.322073: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-05-06 10:02:39.322193: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0
2020-05-06 10:02:39.325903: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-05-06 10:02:39.325946: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 
2020-05-06 10:02:39.325964: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N 
2020-05-06 10:02:39.328011: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 7439 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080, pci bus id: 0000:65:00.0, compute capability: 6.1)
2020-05-06 10:02:41.175170: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
[ASTRI]: inference consumed: 3041 ms

- I think the inference is running on GPU.
- The inference time I mentioned above is only included the time for "session.run()".
- I also attached the inference python script and segmented result for reference.

[reference_model_inference.zip](https://github.com/tensorflow/models/files/4584557/reference_model_inference.zip)
- inference.py: a python script to run a single image inference
- reference_mode.log: the output message during inference
- test.png: segmented results

Thanks.

**Comment by YknZhu** [comment]
Thanks! Could you move the timing inside run(self, image) method, and only wrap around the line of `self.sess.run(...)`? that line is the place where core inference happens.

Also could you try running sess.run() multiple times in a row (e.g. 10)? That would give a better estimate on timing.

**Comment by cclo-astri** [comment]
Hi,

I am not quite understand the meaning of "running sess.run() multiple times in a row". Do you mean perform 10 inferences at one time ?

Thanks.

**Comment by cclo-astri** [comment]
> Thanks! Could you move the timing inside run(self, image) method, and only wrap around the line of `self.sess.run(...)`? that line is the place where core inference happens.

The timing for `self.sess.run(...)` is 2995 ms (while the timing for `model.run(...)` is 3003 ms)

**Comment by cclo-astri** [comment]
> Also could you try running sess.run() multiple times in a row (e.g. 10)? That would give a better estimate on timing.

I just modified the python to loop the `self.sess.run(...)` for 10 times, the timing now is:

> [ASTRI]: [ 0 ] session run consumed: 3008 ms
[ASTRI]: [ 1 ] session run consumed: 197 ms
[ASTRI]: [ 2 ] session run consumed: 197 ms
[ASTRI]: [ 3 ] session run consumed: 196 ms
[ASTRI]: [ 4 ] session run consumed: 198 ms
[ASTRI]: [ 5 ] session run consumed: 196 ms
[ASTRI]: [ 6 ] session run consumed: 197 ms
[ASTRI]: [ 7 ] session run consumed: 197 ms
[ASTRI]: [ 8 ] session run consumed: 197 ms
[ASTRI]: [ 9 ] session run consumed: 197 ms

The time for later 9 inferences seems matched our expected result. 

- Why the 1st frame consumes so much time for inference ? 
- Is there any method to avoid this ? 

Thanks.

**Comment by YknZhu** [comment]
haha that is expected. usually for benchmark models the very first (or the first a few) run could include some initializations and might take longer.

To get stable benchmark timing, it would be better to warm up the model a bit and only measure inference time after warm up period.

**Comment by cclo-astri** [comment]
Hi,

I verified in both python and C++ environment and they have the same behavior (1st inference is slow, but latter inferences become fast). Thanks for your help. 

But I wonder the deeplab model does not support batch inference by default.
- In `_create_input_tensors()` of export_model.py, only a [1, crop_height, crop_width, 3] input tensor created. 
- Error reported when I pass a [batch_size, crop_height, crop_width, 3] tensor to model (where batch_size > 1)

If I want to use deeplab for batch inference for my own data, could I simply change the input tensor shape during exporting my trained model ?

Thanks.

**Comment by YknZhu** [comment]
Haha cool! Happy to know this is helpful.

For batch inference, current export_model doesn’t support it due to preprocessing step: if all of your input images are of the same (known) shape, you can write a batch preprocessing routine to transform it as [batch size, crop_height, crop_width, 3] and replace the current preprocessing step in export_models.py. Then you can also fetch the result after the softmax layer (basically ignore the post_processing step).

**Comment by cclo-astri** [comment]
Hi,

I see. I under-estimated the difficulties of this feature. It's now OK as I can continue my testing with deeplab.

Thanks.

**Comment by KDcx35** [comment]
thank you

**Comment by shokohigol** [comment]
> > Also could you try running sess.run() multiple times in a row (e.g. 10)? That would give a better estimate on timing.
> 
> I just modified the python to loop the `self.sess.run(...)` for 10 times, the timing now is:
> 
> > [ASTRI]: [ 0 ] session run consumed: 3008 ms
> > [ASTRI]: [ 1 ] session run consumed: 197 ms
> > [ASTRI]: [ 2 ] session run consumed: 197 ms
> > [ASTRI]: [ 3 ] session run consumed: 196 ms
> > [ASTRI]: [ 4 ] session run consumed: 198 ms
> > [ASTRI]: [ 5 ] session run consumed: 196 ms
> > [ASTRI]: [ 6 ] session run consumed: 197 ms
> > [ASTRI]: [ 7 ] session run consumed: 197 ms
> > [ASTRI]: [ 8 ] session run consumed: 197 ms
> > [ASTRI]: [ 9 ] session run consumed: 197 ms
> 
> The time for latter 9 inferences seems matched our expected result.
> 
> * Why the 1st frame consumes so much time for inference ?
> * Is there any method to avoid this ?
> 
> Thanks.

This time is correct, but not real time. How to decrease time to 40 ms (use V100 16G or mutli GPU or change in train)

**Comment by cclo-astri** [comment]
Hi @shokohigol , 

As I mentioned at the very beginning, I am only using NVIDIA GTX 1080 (8GB) to perform the experiment. I think there exist many much powerful graphic cards which I think could successfully boost the performance of the inference,

Or if you do want to increase the inference speed without consideration of accuracy, you may try change the backbone network of deeplab from Xception to MobileNet.

For mult-GPU inference, I know nothing about it. If somebody knows how to perform multi-GPU inference with deeplab, please let me know.

Thanks.
